chain/chaintest/block.go

```diff
@@ -42,3 +64,271 @@ func GenerateEmptyExecutedBlocks(
+		if err := tsv.Insert(ctx, chain.HeightKey(metadataManager.HeightPrefix()), binary.BigEndian.AppendUint64(nil, height)); err != nil {
+			return nil, err
+		}
+		if err := tsv.Insert(ctx, chain.TimestampKey(metadataManager.TimestampPrefix()), binary.BigEndian.AppendUint64(nil, uint64(timestamp))); err != nil {
+			return nil, err
+		}
+		if err := tsv.Insert(ctx, chain.FeeKey(metadataManager.FeePrefix()), feeManager.Bytes()); err != nil {
+			return nil, err
+		}
+		tsv.Commit()
+
+		root, err := parentView.GetMerkleRoot(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		parentView, err = parentView.NewView(ctx, merkledb.ViewChanges{
+			MapOps:       ts.ChangedKeys(),
+			ConsumeBytes: true,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		statelessBlock, err := chain.NewStatelessBlock(
+			parentID,
+			timestamp,
+			height,
+			txs,
+			root,
+			nil,
+		)
+		if err != nil {
+			return nil, err
+		}
+		parentID = statelessBlock.GetID()
+
+		blk := chain.NewExecutionBlock(statelessBlock)
+		executionBlocks[i] = blk
+	}
+	return executionBlocks, nil
+}
+
+// BlockBenchmark is a parameterized benchmark. It generates [NumOfBlocks] blocks
+// with [NumOfTxsPerBlock] transactions per block, and then calls
+// Processor.Execute to process the block list [b.N] times.
+type BlockBenchmark struct {
+	MetadataManager chain.MetadataManager
+	BalanceHandler  chain.BalanceHandler
+	RuleFactory     chain.RuleFactory
+	AuthVM          chain.AuthVM
+	TxListGenerator TxListGenerator
+
+	Config                chain.Config
+	AuthVerificationCores int
+
+	// [NumOfBlocks] is set as a hyperparameter to avoid using [b.N] to generate
+	// the number of blocks to run the benchmark on.
+	NumOfBlocks uint64
+	// [NumOfTxsPerBlock] is also equal to the number of unique factories the
+	// benchmark will create.
+	NumOfTxsPerBlock uint64
+}
+
+func (test *BlockBenchmark) Run(ctx context.Context, b *testing.B) {
+	r := require.New(b)
+
+	metrics, err := chain.NewMetrics(prometheus.NewRegistry())
+	r.NoError(err)
+
+	db, err := merkledb.New(
+		ctx,
+		memdb.New(),
+		merkledb.Config{
+			BranchFactor: merkledb.BranchFactor16,
+			Tracer:       trace.Noop,
+		},
+	)
+	r.NoError(err)
+
+	var (
+		processorWorkers          workers.Workers
+		numOfAuthVerificationJobs = 100
+		allocBalance              = 1_000_000_000
+	)
+
+	switch test.AuthVerificationCores {
+	case 0, 1:
+		processorWorkers = workers.NewSerial()
+	default:
+		processorWorkers = workers.NewParallel(
+			test.AuthVerificationCores,
+			numOfAuthVerificationJobs,
+		)
+	}
+
+	factories := []chain.AuthFactory{}
+	customAllocs := []*genesis.CustomAllocation{}
+	for range test.NumOfTxsPerBlock {
+		pk, err := ed25519.GeneratePrivateKey()
+		r.NoError(err)
+
+		factory := auth.NewED25519Factory(pk)
+		factories = append(factories, factory)
+
+		customAllocs = append(customAllocs, &genesis.CustomAllocation{
+			Address: factory.Address(),
+			Balance: uint64(allocBalance),
+		})
+	}
```

Comment:
Should we leave the auth factories behind the generator interface? This is a VM-level detail that will have a non-trivial performance impact.